esi_leap/common/ironic.py

```diff
@@ -23,19 +23,40 @@
 _cached_ironic_client = None
 
 
-def get_ironic_client(context):
+def get_ironic_client(context=None):
     session = ks_loading.load_session_from_conf_options(
         CONF, 'ironic')
     service_auth = ks_loading.load_auth_from_conf_options(CONF, 'ironic')
-    endpoint = ks_loading.load_adapter_from_conf_options(
-        CONF, 'ironic', session=session, auth=service_auth).get_endpoint()
-    user_auth = service_token.ServiceTokenAuthWrapper(
-        user_auth=token_endpoint.Token(endpoint, context.auth_token),
-        service_auth=service_auth)
+
+    # use user context if provided
+    user_auth = None
+    if context:
+        endpoint = ks_loading.load_adapter_from_conf_options(
+            CONF, 'ironic', session=session, auth=service_auth).get_endpoint()
+        user_auth = service_token.ServiceTokenAuthWrapper(
+            user_auth=token_endpoint.Token(endpoint, context.auth_token),
+            service_auth=service_auth)
     sess = ks_loading.load_session_from_conf_options(
-        CONF, 'ironic', auth=user_auth)
+        CONF, 'ironic', auth=user_auth or service_auth)
```

Comment:
When will service_auth be used instead of user_auth?

Reply:
Good question! It should be used when user_auth is None - in other words, when a user context isn't passed in. I was unfamiliar with this pattern, but Ironic uses it: https://opendev.org/openstack/ironic/src/branch/master/ironic/common/neutron.py#L76

Reply:
Got it!